### ISSUE TITLE
Memory issue with sklearn LOO-CV  (on CPU)

### ISSUE THREAD

**Issue by BEPMizusaki** [flatironinstitute/nemos#373]
We came across a memory allocation error when implementing **cross-validation via sklearn (leave-one-out)**, running on **CPU**. This happens after a few folds even if there's 100+GB free memory. See below a minimal code to reproduce it; This particular case is small enough that it runs faster on CPU than on GPU.

Example output:
`JIT session error: Cannot allocate memory`
`Segmentation fault (core dumped)`

This is still ongoing with the latest release of nemos (0.2.3) and **only occurs on fallback to CPU**.
Tested on a number of different linux distributions.
Python versions 3.10.2 ~ 3.12.9
jax 0.4.35 ~ 0.5.3
scikit-learn 1.5.2

It seems to be a jax buffer issue, as it can be "solved" by pipelining **jax_clear_caches()** after every model fit. 
 

 ```
import numpy as np
import nemos as nmo
import sklearn.model_selection as skl
import jax
rnd = np.random.default_rng(seed=1234)

###DATA
n = 200 #Neurons
nf = 10 #number of predictors
nt = 50 #number of epochs

brates = rnd.exponential(10,size=n)
W = rnd.uniform(size=(nf,n))
Mat = rnd.binomial(2,0.7,(nt,nf))-1
Act = np.zeros((nt,n))
for i in range(n):
    lb = brates[i]+np.dot(Mat,W[:,i])
    lb[np.where(lb<0.001)] = 0.01
    Act[:,i]=rnd.poisson(lb)

###MODEL
model1 = nmo.glm.PopulationGLM(regularizer="Ridge")
p1 = np.zeros((nf,n))
p2 =  np.log(np.mean(Act,axis=0))
inn_loop = list(skl.LeaveOneOut().split(np.ones(nt))) 
parameter_grid = {"regularizer_strength": np.logspace(-1, 2, 15)}

crv = skl.GridSearchCV(model1, parameter_grid, cv=inn_loop, error_score='raise', verbose=4)

crv.fit(Mat,Act,init_params= (p1,p2))
```

 **CPU fallback only** - this works normally on GPU.

**Comment by billbrod** [comment]
Hi, thanks for opening this issue with such a clear example. Running your code, I am unable to reproduce this error on our cluster (with CPU only, requesting 50GB of RAM from the job scheduler): the code runs to completion successfully. Using `tracemalloc` to track memory consumption, it looks like the peak memory usage is only ~250MB.

I copied your code into a script with the following contents:
```python
import tracemalloc
import numpy as np
import nemos as nmo
import sklearn.model_selection as skl
import jax
rnd = np.random.default_rng(seed=1234)

###DATA
n = 200 #Neurons
nf = 10 #number of predictors
nt = 50 #number of epochs

brates = rnd.exponential(10,size=n)
W = rnd.uniform(size=(nf,n))
Mat = rnd.binomial(2,0.7,(nt,nf))-1
Act = np.zeros((nt,n))
for i in range(n):
   lb = brates[i]+np.dot(Mat,W[:,i])
   lb[np.where(lb<0.001)] = 0.01
   Act[:,i]=rnd.poisson(lb)

###MODEL
model1 = nmo.glm.PopulationGLM(regularizer="Ridge")
p1 = np.zeros((nf,n))
p2 =  np.log(np.mean(Act,axis=0))
inn_loop = list(skl.LeaveOneOut().split(np.ones(nt)))
parameter_grid = {"regularizer_strength": np.logspace(-1, 2, 15)}

tracemalloc.start()
crv = skl.GridSearchCV(model1, parameter_grid, cv=inn_loop, error_score='raise', verbose=4)
crv.fit(Mat,Act,init_params= (p1,p2))
print(tracemalloc.get_traced_memory())
tracemalloc.stop()
```
running it in a fresh nemos environment (nemos 0.2.3, jax 0.6.2, jaxlib 0.6.2, sklearn 1.6.1, python 3.11.11) with `python test.py` results in output of `(274223603, 274477805)` (tracemalloc reports usage in bytes). Running this with jax 0.5.3 and sklearn 1.5.2 gives the same result (exact memory usage varies, but it never goes about ~250MB).

You say you reproduced this on several linux distributions -- where are you running your code (personal laptop, compute cluster, etc.)?

You say this happens with "CPU fallback only" -- is there a GPU on the machine that hit this error? If so, how are you configuring jax to use the CPU? Or are you running the above code as-is on a machine without a GPU? (I ran this on our cluster, which uses the [rocky](https://rockylinux.org/) linux OS, on a node without GPUs).

It's also clear when I'm running the above code that it's all running in sequence (vs. parallel), because the individual CVs get printed out as they run, each taking 0.3 - 0.5 seconds. Is that also happening for you? ([scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) can run GridSearchCV in parallel, but it's not the default). Most of the jax documentation discussion of memory issues focus on the GPU.

Also, can you show us how you would alter your script to use [`jax.clear_caches()`](https://docs.jax.dev/en/latest/_autosummary/jax.clear_caches.html)? (I'm assuming when you wrote `jax_clear_caches`, that was a typo -- note the underscore vs. period). That way, even if we can't reproduce this error, if anyone else hits this problem we can point them here to solve the issue.

**Comment by billbrod** [comment]
Poking around on the [jax docs](https://docs.jax.dev/en/latest/faq.html#why-do-some-cuda-libraries-fail-to-load-initialize), it looks like there are also a couple of environment variables that might help (though the mention this largely in the context of GPU memory management), by reducing the memory preallocation: setting either `XLA_PYTHON_CLIENT_MEM_FRACTION` to some value lower than its default of 0.75 or `XLA_PYTHON_CLIENT_PREALLOCATE=false`. Do either of those help?

**Comment by BalzaniEdoardo** [comment]
That looks like a memory leak do to buffers not being fully clear on the fallback. 

One interesting thing is that changing regulation strength should not retrigger compilation nor reallocation of memory since it should not be provided as static and it definitely  doesn't change the shape or type of arrays.

Have you tried to log the compilation?

JAX_LOG_COMPILES=1 python script.py

you can also change the allocation policy

XLA_PYTHON_CLIENT_ALLOCATOR=platform

Let me know if this helps clarifying what happens.

Best,
Edoardo
